casper/src/main/scala/coop/rchain/casper/SafetyOracle.scala

```diff
@@ -55,177 +57,29 @@ trait SafetyOracle[F[_]] {
-                  ProtoUtil.getCreatorJustificationAsListUntilGoalInMemory(
-                    blockDag,
-                    blockHash,
-                    b => b == creatorJustificationOrGenesis
-                  )
-                }
-                .pure[F]
-            case None => StreamT.empty[F, BlockHash].pure[F]
-          }
-
-        def neverEventuallySeeDisagreement(
-            first: Validator,
-            second: Validator
-        ): F[Boolean] =
-          (for {
-            firstLatestBlock <- OptionT(blockDag.latestMessage(first))
-            secondLatestOfFirstLatestHash <- OptionT.fromOption[F](
-                                              firstLatestBlock.justifications
-                                                .find {
-                                                  case Justification(validator, _) =>
-                                                    validator == second
-                                                }
-                                                .map(_.latestBlockHash)
-                                            )
-            secondLatestOfFirstLatest <- OptionT(blockDag.lookup(secondLatestOfFirstLatestHash))
-            potentialDisagreements <- OptionT.liftF(
-                                       filterChildren(secondLatestOfFirstLatest, second)
-                                     )
-            // TODO: Implement forallM on StreamT
-            result <- OptionT.liftF(potentialDisagreements.toList.flatMap(_.forallM {
-                       potentialDisagreement =>
-                         computeCompatibility(blockDag, candidateMetadata, potentialDisagreement)
-                     }))
-          } yield result).fold(false)(identity)
-
-        def computeAgreementGraphEdges: F[List[(Validator, Validator)]] =
-          (for {
-            x <- agreeingValidatorToWeight.keys
-            y <- agreeingValidatorToWeight.keys
-            if x.toString > y.toString // TODO: Order ByteString
-          } yield (x, y)).toList.filterA {
-            case (first: Validator, second: Validator) =>
-              neverEventuallySeeDisagreement(first, second) &&^ neverEventuallySeeDisagreement(
-                second,
-                first
-              )
-          }
-
-        computeAgreementGraphEdges.map { edges =>
-          Clique.findMaximumCliqueByWeight[Validator](edges, agreeingValidatorToWeight)
-        }
-      }
-
-      private def computeCompatibility(
-          blockDag: BlockDagRepresentation[F],
-          candidateMetadata: BlockMetadata,
-          targetBlockHash: BlockHash
-      ): F[Boolean] =
-        isInMainChain(blockDag, candidateMetadata, targetBlockHash)
+      CliqueOracle.normalizedFaultTolerance(
+        blockDag,
+        candidateBlockHash,
+        (h: BlockHash) => blockDag.lookupUnsafe(h).map(_.weightMap),
+        (h: BlockHash) => blockDag.lookupUnsafe(h).map(_.blockNum),
```

Comment:
I wonder - is it better to created BlockDagRepresentation that will be DagReader[F, BlockMetadata] and get rid of `F[_]` in CliqueOracle and some other places? Or this more fine grained access to F[_] is better?

Reply:
I was thinking for a while and my guess is that better to leave for input functions of CliqueOracle possibility to not be pure, but inside caller decide if we need to get e.g. `DagReader[F, BlockMetadata]` and then pass in something like `meta.weightMap.pure[F]`, or acquire values from `F[_]` for each function separately.